### DISCUSSION TITLE
How to develop VSCode?

### DISCUSSION THREAD

**Discussion by bilogic** [coder/code-server#6884] [Q&A]
I run a headless Ubuntu and like to ask how you guys efficiently develop `code-server`, which in a way, is equivalent to developing `VSCode`.

Do you use `code-server` to edit the code of `VSCode`? There are some steps in the readme, I followed https://coder.com/docs/code-server/CONTRIBUTING#development-workflow but nothing seems to be listening on 8080 after the `yarn watch` command.

**Comment by code-asher** [comment]
I do not use code-server to edit code-server or VS Code, but it should work as well as any other editor.

What do you see at the end of the `yarn watch` command?  You should see something like this:

```
[code-server] [2024-07-08T18:53:25.663Z] info  code-server 0.0.0 development
[code-server] [2024-07-08T18:53:25.664Z] info  Using user-data-dir /home/coder/.local/share/code-server
[code-server] [2024-07-08T18:53:25.677Z] info  Using config file /home/coder/.config/code-server/config.yaml
[code-server] [2024-07-08T18:53:25.677Z] info  HTTP server listening on http://127.0.0.1:8081/
[code-server] [2024-07-08T18:53:25.677Z] info    - Authentication is enabled
[code-server] [2024-07-08T18:53:25.677Z] info      - Using password from /home/coder/.config/code-server/config.yaml
[code-server] [2024-07-08T18:53:25.677Z] info    - Not serving HTTPS
[code-server] [2024-07-08T18:53:25.677Z] info  Session server listening on /home/coder/.local/share/code-server/code-server-ipc.sock
```

In my case, it is listening on `8081`.  `8080` is the default, but I have it changed in the `config.yaml`.

**Reply by code-asher** [reply]
Also keep in mind that loading the development version in the browser is veeeeeeeery slow.

**Reply by bilogic** [reply]
```
[~/code-server]# yarn
yarn install v1.22.22
[1/5] Validating package.json...
[2/5] Resolving packages...
success Already up-to-date.
$ ./ci/dev/postinstall.sh
Installing dependencies for /root/code-server/test
yarn install v1.22.22
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.30s.
Installing dependencies for /root/code-server/test/e2e/extensions/test-extension
yarn install v1.22.22
[1/5] Validating package.json...
warning code-server-extension@0.0.1: The engine "vscode" appears to be invalid.
```

I get this when I run yarn `warning code-server-extension@0.0.1: The engine "vscode" appears to be invalid.`. Is it a problem?

**Reply by bilogic** [reply]
```
[code-server] [2024-07-09T03:11:35.091Z] info  code-server 0.0.0 development
[code-server] [2024-07-09T03:11:35.092Z] info  Using user-data-dir /root/.local/share/code-server
[code-server] [2024-07-09T03:11:35.110Z] info  Using config file /root/.config/code-server/config.yaml
[code-server] [2024-07-09T03:11:35.110Z] info  HTTP server listening on http://0.0.0.0:8080/
[code-server] [2024-07-09T03:11:35.110Z] info    - Authentication is enabled
[code-server] [2024-07-09T03:11:35.110Z] info      - Using password from /root/.config/code-server/config.yaml
[code-server] [2024-07-09T03:11:35.110Z] info    - Not serving HTTPS
[code-server] [2024-07-09T03:11:35.110Z] info  Session server listening on /root/.local/share/code-server/code-server-ipc.sock
```

I got it to listen on `:8080`, but Chrome dev tools console has many errors and it eventually stopped at a totally blank page

**Reply by bilogic** [reply]
```
git submodule update --init
quilt push -a
yarn install
yarn build
VERSION=0.0.0 yarn build:vscode
yarn release
```

I decided to do a full compilation, but `yarn build:vscode` takes over an hour on a 24 core 64GB xeon... how is that possible? And I can't make it show more verbose logs.

**Reply by bilogic** [reply]
I was able to get VSCode to compile in under 10 minutes, it would appear there is an issue with the `gulp` that keeps running for over an hour.

**Reply by bilogic** [reply]
I'm really hoping to move on this https://github.com/coder/code-server/issues/6496 thanks

**Reply by code-asher** [reply]
In case anyone else is interested in this thread, we are talking about this more here: https://github.com/coder/code-server/issues/6891

**Comment by benz0li** [comment]
@bilogic I use image `glcr.b-data.ch/jupyterlab/python/base:latest-devtools` from b-data's/my [JupyterLab Python docker stack](https://github.com/b-data/jupyterlab-python-docker-stack) and use `code-server` as code editor. This image includes the requirements according to

* [coder/code-server: Docs > Contributing](https://github.com/coder/code-server/blob/main/docs/CONTRIBUTING.md)
  * [microsoft/vscode: Wiki > How to Contribute > Contributing to Visual Studio Code](https://github.com/microsoft/vscode/wiki/How-to-Contribute)
